File: tests/test_sld.py
"""Tests for the SLD (Symmetric Logarithmic Derivative) module."""

import pytest
import numpy as np
import mpmath as mp

from lmg_qfi import (
    compute_sld_matrix,
    sld_expectation,
    sld_squared_expectation,
    compute_sld_info,
    quantum_fisher_information_mp,
    SLDInformation,
    create_spin_xyz_operators,
)

np.set_printoptions(precision=4, suppress=True)


def convert_mpmatrix_to_numpy(mp_math, precision=15):
    """Convert mpmath matrix to numpy array."""
    with mp.workdps(precision):
        np_mat = np.array(
            [[complex(x.real, x.imag) for x in row] for row in mp_math.tolist()],
            dtype=np.complex128
        )
    return np_mat


class TestSLDMatrix:
    """Tests for compute_sld_matrix."""

    @pytest.mark.parametrize("n", [2, 5, 10])
    def test_sld_matrix_hermitian(self, n):
        """L = 2(|∂ψ⟩⟨ψ| + |ψ⟩⟨∂ψ|) must be Hermitian."""
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.1')

        L = compute_sld_matrix(dket, ket)
        L_np = convert_mpmatrix_to_numpy(L)

        assert np.allclose(L_np, L_np.T.conj(), atol=1e-12), "SLD matrix is not Hermitian"

    @pytest.mark.parametrize("n", [2, 5])
    def test_sld_equation(self, n):
        """(ρL + Lρ)/2 must equal ∂ρ for normalized states."""
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.3')

        L = compute_sld_matrix(dket, ket)
        rho = ket * ket.transpose_conj()
        d_rho = dket * ket.transpose_conj() + ket * dket.transpose_conj()

        lhs = (rho * L + L * rho) * mp.mpf('0.5')
        lhs_np = convert_mpmatrix_to_numpy(lhs)
        d_rho_np = convert_mpmatrix_to_numpy(d_rho)

        assert np.allclose(lhs_np, d_rho_np, atol=1e-12), "SLD equation not satisfied"

    def test_sld_matrix_shape(self):
        """SLD matrix has shape (n+1, n+1)."""
        n = 4
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.1')

        L = compute_sld_matrix(dket, ket)

        assert L.rows == n + 1
        assert L.cols == n + 1

    def test_sld_matrix_zero_derivative(self):
        """SLD is zero when derivative ket is zero."""
        n = 3
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)

        L = compute_sld_matrix(dket, ket)
        L_np = convert_mpmatrix_to_numpy(L)

        assert np.allclose(L_np, np.zeros((n + 1, n + 1)), atol=1e-12)


class TestSLDExpectation:
    """Tests for sld_expectation (⟨L⟩)."""

    @pytest.mark.parametrize("n", [2, 5, 10])
    def test_sld_expectation_zero_orthogonal(self, n):
        """⟨L⟩ = 0 when derivative is orthogonal to state."""
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.5')  # orthogonal to ket[0]

        result = sld_expectation(dket, ket)

        assert mp.fabs(result) < 1e-12

    def test_sld_expectation_purely_imaginary_overlap(self):
        """⟨L⟩ = 0 when ⟨ψ|∂ψ⟩ is purely imaginary (normalized state)."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        # Purely imaginary overlap: dket = i * c * ket
        dket = mp.matrix([[mp.mpc(0, '0.3')], [mp.mpf('0.0')]])

        result = sld_expectation(dket, ket)

        # Re(⟨ψ|∂ψ⟩) = Re(0.3i) = 0, so ⟨L⟩ = 0
        assert mp.fabs(result) < 1e-12

    def test_sld_expectation_formula(self):
        """⟨L⟩ = 4 Re(⟨ψ|∂ψ⟩) matches direct computation."""
        ket = mp.matrix([[mp.mpf('0.6')], [mp.mpf('0.8')]])
        dket = mp.matrix([[mp.mpf('0.1')], [mp.mpf('0.2')]])

        result = sld_expectation(dket, ket)
        expected = 4 * mp.re((ket.transpose_conj() * dket)[0, 0])

        assert mp.fabs(result - expected) < 1e-12


class TestSLDSquaredExpectation:
    """Tests for sld_squared_expectation (⟨L²⟩)."""

    @pytest.mark.parametrize("n", [2, 5, 10])
    def test_sld_squared_matches_qfi(self, n):
        """⟨L²⟩ must equal the finite-difference QFI for a normalized state."""
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.1')

        L2 = sld_squared_expectation(dket, ket)
        qfi = quantum_fisher_information_mp(dket, ket)

        assert mp.fabs(L2 - qfi) < 1e-12, f"⟨L²⟩={L2} != QFI={qfi}"

    def test_sld_squared_non_negative(self):
        """⟨L²⟩ must be non-negative."""
        ket = mp.matrix([[mp.mpf('0.6')], [mp.mpf('0.8')]])
        dket = mp.matrix([[mp.mpf('0.1')], [mp.mpf('-0.2')]])

        result = sld_squared_expectation(dket, ket)

        assert result >= 0

    def test_sld_squared_zero_for_zero_derivative(self):
        """⟨L²⟩ = 0 when derivative ket is zero."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        dket = mp.matrix([[mp.mpf('0.0')], [mp.mpf('0.0')]])

        result = sld_squared_expectation(dket, ket)

        assert mp.fabs(result) < 1e-12

    def test_sld_squared_via_full_matrix(self):
        """⟨L²⟩ matches Tr(ρ L²) computed from the full SLD matrix."""
        n = 3
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.2')
        dket[2] = mp.mpf('0.1')

        L = compute_sld_matrix(dket, ket)
        rho = ket * ket.transpose_conj()
        # Tr(ρ L²) = ⟨ψ|L²|ψ⟩ for pure state
        tr_rho_L2 = mp.re((ket.transpose_conj() * L * L * ket)[0, 0])
        L2_fast = sld_squared_expectation(dket, ket)

        assert mp.fabs(tr_rho_L2 - L2_fast) < 1e-12

    @pytest.mark.parametrize("n", [2, 5])
    def test_sld_squared_global_phase_invariance(self, n):
        """⟨L²⟩ is invariant under global phase on ket."""
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0') / mp.sqrt(2)
        ket[1] = mp.mpf('1.0') / mp.sqrt(2)
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.1')

        L2_original = sld_squared_expectation(dket, ket)

        phase = mp.exp(mp.j * mp.pi / 3)
        L2_phased = sld_squared_expectation(phase * dket, phase * ket)

        assert mp.fabs(L2_original - L2_phased) < 1e-12


class TestComputeSLDInfo:
    """Tests for compute_sld_info."""

    def test_returns_sld_information(self):
        """compute_sld_info returns an SLDInformation instance."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        dket = mp.matrix([[mp.mpf('0.0')], [mp.mpf('0.1')]])

        result = compute_sld_info(dket, ket, N=2, time=5)

        assert isinstance(result, SLDInformation)

    def test_time_field(self):
        """time field matches the input."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        dket = mp.matrix([[mp.mpf('0.0')], [mp.mpf('0.1')]])

        result = compute_sld_info(dket, ket, N=2, time=7)

        assert result.time == 7

    def test_qfi_from_sld_normalization(self):
        """qfi_from_sld = L_squared_expectation / (N² t²)."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        dket = mp.matrix([[mp.mpf('0.0')], [mp.mpf('0.1')]])
        N, t = 2, 5

        result = compute_sld_info(dket, ket, N=N, time=t)

        expected = result.L_squared_expectation / (N ** 2 * t ** 2)
        assert abs(result.qfi_from_sld - expected) < 1e-12

    def test_l_expectation_near_zero(self):
        """L_expectation is ~0 for a normalized state with orthogonal derivative."""
        ket = mp.matrix([[mp.mpf('1.0')], [mp.mpf('0.0')]])
        dket = mp.matrix([[mp.mpf('0.0')], [mp.mpf('0.3')]])

        result = compute_sld_info(dket, ket, N=2, time=3)

        assert abs(result.L_expectation) < 1e-12

    @pytest.mark.parametrize("n", [2, 5, 10])
    def test_qfi_from_sld_matches_finite_difference(self, n):
        """Normalized QFI from SLD matches finite-difference QFI / (N² t²)."""
        N = n
        t = 4
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf('1.0')
        dket = mp.zeros(n + 1, 1)
        dket[1] = mp.mpf('0.15')

        result = compute_sld_info(dket, ket, N=N, time=t)
        qfi_fd = float(quantum_fisher_information_mp(dket, ket))
        qfi_normalized = qfi_fd / (N ** 2 * t ** 2)

        assert abs(result.qfi_from_sld - qfi_normalized) < 1e-12


class TestSLDExpectationExtended:
    """Additional tests for sld_expectation edge cases."""

    def test_sld_expectation_nonzero_for_real_overlap(self):
        """⟨L⟩ = 4*Re(⟨ψ|∂ψ⟩) != 0 when derivative has real component along ket."""
        ket = mp.matrix([[mp.mpf("1.0")], [mp.mpf("0.0")]])
        dket = mp.matrix([[mp.mpf("0.1")], [mp.mpf("0.0")]])

        result = sld_expectation(dket, ket)

        # ⟨ψ|∂ψ⟩ = 0.1 (real) → ⟨L⟩ = 4 * 0.1 = 0.4
        assert mp.fabs(result - mp.mpf("0.4")) < 1e-12

    @pytest.mark.parametrize("n", [2, 5])
    def test_sld_expectation_zero_for_sy_generator(self, n):
        """⟨L⟩ = 0 for Sy-generated derivative (purely imaginary overlap)."""
        Sz, Sx, Sy = create_spin_xyz_operators(n)
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf("1.0")
        dket = mp.j * Sy * ket

        result = sld_expectation(dket, ket)

        assert float(mp.fabs(result)) < 1e-10

    def test_sld_expectation_superposition(self):
        """⟨L⟩ = 4*Re(⟨ψ|∂ψ⟩) for equal superposition state."""
        ket = mp.matrix(
            [[mp.mpf("1.0") / mp.sqrt(2)], [mp.mpf("1.0") / mp.sqrt(2)]]
        )
        dket = mp.matrix([[mp.mpc("0.1", "0.05")], [mp.mpc("-0.1", "0.05")]])

        result = sld_expectation(dket, ket)
        expected = 4 * mp.re((ket.transpose_conj() * dket)[0, 0])

        assert mp.fabs(result - expected) < 1e-12


class TestSLDSquaredExtended:
    """Additional tests for sld_squared_expectation."""

    @pytest.mark.parametrize("n", [2, 5])
    def test_sld_squared_matches_qfi_for_sx_generator(self, n):
        """⟨L²⟩ = QFI for Sx-generated derivative (imaginary overlap)."""
        Sz, Sx, Sy = create_spin_xyz_operators(n)
        ket = mp.zeros(n + 1, 1)
        ket[0] = mp.mpf("1.0")
        dket = mp.j * Sx * ket

        alpha = (ket.transpose_conj() * dket)[0, 0]
        assert mp.fabs(mp.re(alpha)) < 1e-10  # verify premise

        sld_sq = sld_squared_expectation(dket, ket)
        qfi = quantum_fisher_information_mp(dket, ket)

        assert mp.fabs(sld_sq - qfi) < 1e-10

    def test_sld_squared_known_value_orthogonal(self):
        """⟨L²⟩ = 4 for |ψ⟩=|0⟩ and |∂ψ⟩=|1⟩."""
        ket = mp.matrix([[mp.mpf("1")], [mp.mpf("0")]])
        dket = mp.matrix([[mp.mpf("0")], [mp.mpf("1")]])

        result = sld_squared_expectation(dket, ket)

        assert mp.fabs(result - 4) < 1e-12


class TestSLDInformation:
    """Tests for SLDInformation dataclass."""

    def test_sld_information_creation(self):
        """Test creating SLDInformation with all fields."""
        info = SLDInformation(
            time=10,
            L_expectation=0.001,
            L_squared_expectation=1.5,
            qfi_from_sld=0.015,
        )

        assert info.time == 10
        assert info.L_expectation == 0.001
        assert info.L_squared_expectation == 1.5
        assert info.qfi_from_sld == 0.015

    def test_sld_information_all_fields_accessible(self):
        """All expected fields are present on SLDInformation."""
        info = SLDInformation(
            time=1,
            L_expectation=0.0,
            L_squared_expectation=0.0,
            qfi_from_sld=0.0,
        )

        assert hasattr(info, "time")
        assert hasattr(info, "L_expectation")
        assert hasattr(info, "L_squared_expectation")
        assert hasattr(info, "qfi_from_sld")

    def test_sld_information_float_compatible(self):
        """Numeric fields are float-compatible."""
        info = SLDInformation(
            time=50,
            L_expectation=1e-5,
            L_squared_expectation=3.14,
            qfi_from_sld=0.0628,
        )

        assert isinstance(float(info.L_expectation), float)
        assert isinstance(float(info.L_squared_expectation), float)
        assert isinstance(float(info.qfi_from_sld), float)

    def test_sld_information_zero_values(self):
        """SLDInformation with zero values (stationary state)."""
        info = SLDInformation(
            time=100,
            L_expectation=0.0,
            L_squared_expectation=0.0,
            qfi_from_sld=0.0,
        )

        assert info.L_expectation == 0.0
        assert info.L_squared_expectation == 0.0
        assert info.qfi_from_sld == 0.0

    def test_sld_information_roundtrip_with_compute_sld_info(self):
        """SLDInformation produced by compute_sld_info has correct field types."""
        ket = mp.matrix([[mp.mpf("1.0")], [mp.mpf("0.0")]])
        dket = mp.matrix([[mp.mpf("0.0")], [mp.mpf("0.2")]])

        info = compute_sld_info(dket, ket, N=3, time=4)

        assert isinstance(info, SLDInformation)
        assert isinstance(info.time, (int, float))
        assert isinstance(info.L_expectation, float)
        assert isinstance(info.L_squared_expectation, float)
        assert isinstance(info.qfi_from_sld, float)

File: fortran/tests/test_sld.f90
program test_sld
  use lmg_constants, only: dp, CI, CZERO, CONE
  use lmg_types, only: SLDInformation_type
  use lmg_sld, only: compute_sld_matrix, sld_expectation, sld_squared_expectation, compute_sld_info
  use lmg_qfi_mod, only: quantum_fisher_information
  use test_utils
  implicit none

  call test_suite_start('lmg_sld')

  call test_sld_matrix_hermitian()
  call test_sld_equation()
  call test_sld_matrix_zero_derivative()
  call test_sld_expectation_orthogonal()
  call test_sld_expectation_formula()
  call test_sld_expectation_nonzero_real()
  call test_sld_squared_matches_qfi()
  call test_sld_squared_non_negative()
  call test_sld_squared_zero_derivative()
  call test_sld_squared_orthogonal()
  call test_sld_squared_phase_invariance()
  call test_compute_sld_info_time()
  call test_compute_sld_info_normalization()
  call test_compute_sld_info_l_exp_near_zero()

  call test_suite_end()
  if (tests_failed > 0) stop 1

contains

  subroutine test_sld_matrix_hermitian()
    integer, parameter :: dim = 6
    complex(dp) :: ket(dim), dket(dim), L(dim, dim)

    call test_start('SLD matrix Hermitian')

    ket = CZERO; ket(1) = CONE
    dket = CZERO; dket(2) = (0.1_dp, 0.0_dp)

    call compute_sld_matrix(dket, ket, dim, L)
    call assert_matrix_hermitian(L, dim, 1e-12_dp, 'L Hermitian')
    call test_pass()
  end subroutine

  subroutine test_sld_equation()
    integer, parameter :: dim = 3
    complex(dp) :: ket(dim), dket(dim), L(dim, dim)
    complex(dp) :: rho(dim,dim), d_rho(dim,dim), lhs(dim,dim)
    integer :: i, j

    call test_start('SLD equation')

    ket = CZERO; ket(1) = CONE
    dket = CZERO; dket(2) = (0.3_dp, 0.0_dp)

    call compute_sld_matrix(dket, ket, dim, L)

    ! rho = |ket><ket|
    do j = 1, dim
      do i = 1, dim
        rho(i,j) = ket(i) * conjg(ket(j))
        d_rho(i,j) = dket(i) * conjg(ket(j)) + ket(i) * conjg(dket(j))
      end do
    end do

    ! lhs = (rho*L + L*rho) / 2
    lhs = 0.5_dp * (matmul(rho, L) + matmul(L, rho))
    call assert_matrix_close(lhs, d_rho, dim, 1e-12_dp, 'SLD equation')
    call test_pass()
  end subroutine

  subroutine test_sld_matrix_zero_derivative()
    integer, parameter :: dim = 4
    complex(dp) :: ket(dim), dket(dim), L(dim, dim), zero_mat(dim, dim)

    call test_start('SLD zero for zero deriv')

    ket = CZERO; ket(1) = CONE
    dket = CZERO
    zero_mat = CZERO

    call compute_sld_matrix(dket, ket, dim, L)
    call assert_matrix_close(L, zero_mat, dim, 1e-12_dp, 'L=0')
    call test_pass()
  end subroutine

  subroutine test_sld_expectation_orthogonal()
    integer, parameter :: dim = 6
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: result

    call test_start('<L>=0 orthogonal')

    ket = CZERO; ket(1) = CONE
    dket = CZERO; dket(2) = (0.5_dp, 0.0_dp)

    result = sld_expectation(dket, ket, dim)
    call assert_close(abs(result), 0.0_dp, 1e-12_dp, '<L>=0')
    call test_pass()
  end subroutine

  subroutine test_sld_expectation_formula()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim), alpha
    real(dp) :: result, expected

    call test_start('<L> = 4*Re(<ket|dket>)')

    ket = [(0.6_dp, 0.0_dp), (0.8_dp, 0.0_dp)]
    dket = [(0.1_dp, 0.0_dp), (0.2_dp, 0.0_dp)]

    result = sld_expectation(dket, ket, dim)
    alpha = dot_product(ket, dket)
    expected = 4.0_dp * real(alpha, dp)

    call assert_close(result, expected, 1e-12_dp, '<L> formula')
    call test_pass()
  end subroutine

  subroutine test_sld_expectation_nonzero_real()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: result

    call test_start('<L>=0.4 real overlap')

    ket = [CONE, CZERO]
    dket = [(0.1_dp, 0.0_dp), CZERO]

    result = sld_expectation(dket, ket, dim)
    call assert_close(result, 0.4_dp, 1e-12_dp, '<L>=0.4')
    call test_pass()
  end subroutine

  subroutine test_sld_squared_matches_qfi()
    integer, parameter :: dim = 6
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: L2, qfi

    call test_start('<L^2> = QFI')

    ket = CZERO; ket(1) = CONE
    dket = CZERO; dket(2) = (0.1_dp, 0.0_dp)

    L2 = sld_squared_expectation(dket, ket, dim)
    qfi = quantum_fisher_information(dket, ket, dim)

    call assert_close(L2, qfi, 1e-12_dp, '<L^2>=QFI')
    call test_pass()
  end subroutine

  subroutine test_sld_squared_non_negative()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: result

    call test_start('<L^2> >= 0')

    ket = [(0.6_dp, 0.0_dp), (0.8_dp, 0.0_dp)]
    dket = [(0.1_dp, 0.0_dp), (-0.2_dp, 0.0_dp)]

    result = sld_squared_expectation(dket, ket, dim)
    call assert_true(result >= -1e-12_dp, '<L^2> non-negative')
    call test_pass()
  end subroutine

  subroutine test_sld_squared_zero_derivative()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: result

    call test_start('<L^2>=0 zero deriv')

    ket = [CONE, CZERO]
    dket = CZERO

    result = sld_squared_expectation(dket, ket, dim)
    call assert_close(abs(result), 0.0_dp, 1e-12_dp, '<L^2>=0')
    call test_pass()
  end subroutine

  subroutine test_sld_squared_orthogonal()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    real(dp) :: result

    call test_start('<L^2>=4 orthogonal')

    ket = [CONE, CZERO]
    dket = [CZERO, CONE]

    result = sld_squared_expectation(dket, ket, dim)
    call assert_close(result, 4.0_dp, 1e-12_dp, '<L^2>=4')
    call test_pass()
  end subroutine

  subroutine test_sld_squared_phase_invariance()
    integer, parameter :: dim = 3
    complex(dp) :: ket(dim), dket(dim), phase
    real(dp) :: L2_orig, L2_phased

    call test_start('<L^2> phase invariant')

    ket = CZERO
    ket(1) = cmplx(1.0_dp / sqrt(2.0_dp), 0.0_dp, dp)
    ket(2) = cmplx(1.0_dp / sqrt(2.0_dp), 0.0_dp, dp)
    dket = CZERO
    dket(2) = (0.1_dp, 0.0_dp)

    L2_orig = sld_squared_expectation(dket, ket, dim)

    phase = exp(CI * 1.047197551196598_dp)  ! pi/3
    L2_phased = sld_squared_expectation(phase * dket, phase * ket, dim)

    call assert_close(L2_orig, L2_phased, 1e-12_dp, '<L^2> invariant')
    call test_pass()
  end subroutine

  subroutine test_compute_sld_info_time()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    type(SLDInformation_type) :: info

    call test_start('sld_info time field')

    ket = [CONE, CZERO]
    dket = [CZERO, (0.1_dp, 0.0_dp)]

    call compute_sld_info(dket, ket, dim, 2, 7, info)
    call assert_true(info%time == 7, 'time=7')
    call test_pass()
  end subroutine

  subroutine test_compute_sld_info_normalization()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    type(SLDInformation_type) :: info
    real(dp) :: expected

    call test_start('sld_info normalization')

    ket = [CONE, CZERO]
    dket = [CZERO, (0.1_dp, 0.0_dp)]

    call compute_sld_info(dket, ket, dim, 2, 5, info)
    expected = info%L_squared_expectation / (4.0_dp * 25.0_dp)  ! N=2, t=5
    call assert_close(info%qfi_from_sld, expected, 1e-12_dp, 'qfi_from_sld')
    call test_pass()
  end subroutine

  subroutine test_compute_sld_info_l_exp_near_zero()
    integer, parameter :: dim = 2
    complex(dp) :: ket(dim), dket(dim)
    type(SLDInformation_type) :: info

    call test_start('sld_info <L>~0')

    ket = [CONE, CZERO]
    dket = [CZERO, (0.3_dp, 0.0_dp)]

    call compute_sld_info(dket, ket, dim, 2, 3, info)
    call assert_close(abs(info%L_expectation), 0.0_dp, 1e-12_dp, '<L>~0')
    call test_pass()
  end subroutine

end program test_sld
